### PR TITLE
ANW-1564 further refactoring of admin-defined subrecord requirements

### PR DESCRIPTION
**Pull request by quoideneuf**
This is a follow up / continuation of https://github.com/archivesspace/archivesspace/pull/2703 and a refactor of https://github.com/archivesspace/archivesspace/pull/1073. 

What's changed:

The API endpoint `repositories/:repo_id/required_fields/:record_type` now returns an object like this:

```json
{
  "lock_version": 0,
  "record_type": "agent_person",
  "subrecord_requirements": [
    {
      "property": "agent_other_agency_codes",
      "required_fields": [
        "agency_code_type"
      ],
      "required": false
    },
   {
      "property": "metadata_rights_declarations",
      "record_type": "metadata_rights_declaration",
      "required": true,
      "required_fields": [
        "license",
        "descriptive_note"
      ]
    },
    {
      "property": "agent_resources",
      "required_fields": [
        "file_version_xlink_show_attribute"
      ],
      "required": false
    }
  ],
  "jsonmodel_type": "required_fields",
  "uri": "/repositories/2/required_fields/agent_person"
}
```
The `record_type` is optional, since legacy datasets will have blobs that look like this:

```json
{
  "record_type": "agent_person",
  "lock_version": 1,
  "required": {
    "agent_type": "agent_person",
    "agent_other_agency_codes": [
      {
        "agency_code_type": "REQ"
      }
    ],
    "metadata_rights_declarations": [
      {
        "xlink_role_attribute": "REQ",
        "xlink_arcrole_attribute": "REQ",
        "last_verified_date": "REQ"
      }
    ],
    "agent_alternate_sets": [
      {
        "descriptive_note": "REQ"
      }
    ],
    "agent_resources": [
      {
        "file_version_xlink_show_attribute": "REQ",
        "publish": false
      }
    ],
    "is_linked_to_published_record": false,
    "system_generated": false,
    "publish": false,
    "is_slug_auto": false
  },
  "jsonmodel_type": "required_fields",
  "uri": "/repositories/2/required_fields/agent_person"
}
```
We probably could have inferred the record type of these legacy entries, since in practice a property and record_type have a 1:1 relationship. But since this is not absolute and the schema structures allow for properties that can take more than one record type, we decided to implement a requirements model in the frontend that can decide if a field is required in cases where: a) the requirements are missing a type; b) the caller doesn't know the type; c) the caller doesn't know the property. It seems to work, since in practice the only area where admin-defined requirements has been implemented  is in Agent records, and these only have properties that allow a single sub record type.

This points to a more general issue with properties and record types that can be matched to them. There seem to be many places in ArchivesSpace, especially in views, where there is a non-explicit dependency on only one type being found as a property (sub record) of another. There is also a lot of repetition in views, repeated property and record type names, which could perhaps be eliminated.

### CODE REVIEW

**Comment by quoideneuf** [comment]
Below are the schema that currently have properties allowing more than one record type (excluding properties that contain references to JSONModel URIs). Most of these are non top-level schema, so it would be irrelevant for the record requirements feature as it is now. The main groups that are top-level properties are agent relationships and notes. 

[agent_topic.notes](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/agent_topic.rb#L14-L17)

[archival_object.notes](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/archival_object.rb#L95-L101)

[advanced_query.query](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/advanced_query.rb#L8)

[digital_object.notes](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/digital_object.rb#L56-L60)

[agent_corporate_entity.related_agents](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/agent_corporate_entity.rb#L23-L34)

[note_multipart.subnotes](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/note_multipart.rb#L20-L26)

[agent_gender.notes](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/agent_gender.rb#L14-L18)

[resource.notes](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/resource.rb#L123-L129)

[agent_person.related_agents](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/agent_person.rb#L29-L39)

[note_structure_or_genealogy.subnotes](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/note_structure_or_genealogy.rb#L10-L16)

[agent_contact.notes](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/agent_contact.rb#L25-L29)

[agent_place.notes](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/agent_place.rb#L16-L20)

[boolean_query.subqueries](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/boolean_query.rb#L9)

[digital_object_component.notes](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/digital_object_component.rb#L49-L53)

[note_legal_status.subnotes](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/note_legal_status.rb#L10-L14)

[agent_function.notes](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/agent_function.rb#L9-L13)

[note_general_context.subnotes](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/note_general_context.rb#L10-L16)

[agent_occupation.subnotes](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/agent_occupation.rb#L9-L13)

[agent_family.related_agents](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/agent_family.rb#L23-L32)

[note_bioghist.subnotes](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/note_bioghist.rb#L10-L19)

[note_mandate.subnotes](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/note_mandate.rb#L10-L14)

[abstract_agent.notes](https://github.com/archivesspace/archivesspace/blob/master/common/schemas/abstract_agent.rb#L131-L138)

**Review by mark-cooper — APPROVED** [review]
I tried this and everything I attempted worked great. On the code level it's a really nice improvement -- there are some complicated looking parts that are hard to follow (at least for me) without specifically tracing / prying through it but a) that's ArchivesSpace 😄 b) it's well documented, and c) the overall complexity has been significantly reduced compared with the previous implementation 👍
